sourcegraph/sourcegraph
# search: support draining Zoekt instances

This change lets us set an ENV (`INDEXED_SEARCH_DRAIN_SERVERS`) 
on frontend with the Zoekt endpoints we want to drain.
As a consequence all repos indexed by those instances will be
assigned to other instances. Search still targets all instances. After
an instance has been fully drained, we can remove it without user
impact.

Here is how the process would look like:

1. Use #61832 to get a list of endpoints
2. Decide which endpoints to drain. Then set ENV on frontend and restart.
3. Monitor Zoekt dashboard to see when all indexes have been migrated. Depending on the number and size of repos, this can take minutes,  hours or even days.
4. Remove empty instance
5. Unset ENV

Test plan:
- New unit test
- manual testing

As soon as the ENV is set, indexes are assigned to the "green" instance. Once the repos have been indexed on "green", they are dropped by "yellow". As one can see in the graph, there is a short period during which repos might appear on both instances. We already have a [deduper](https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph@405150fba2710a58e88120f412503366e0af8706/-/blob/internal/search/backend/horizontal.go?L60-62) in place which takes care of that.

<img width="1111" alt="Screenshot 2024-04-12 at 14 38 34" src="https://github.com/sourcegraph/sourcegraph/assets/26413131/d908638c-e378-428d-8532-04c2fbf75001">
